### PR TITLE
Add HTTP body streaming

### PR DESCRIPTION
**Pull request by legionth**
Added a possibility to stream the entire body via ReactPHP [stream](https://github.com/reactphp/stream) but still using the needed `StreamInterface` defined by the PSR-7 specification for response objects.

The body which will generated by the user of the server, can use an ReactPHP stream to stream the entire body to the client. Check out the example in the `examples` folder.

Most of the functions from the `StreamInterface` aren't needed for further operations of the `HttpServer`, that's why there so many dummy return values for these.

### CODE REVIEW

File: examples/streamServer.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+require __DIR__ . '/../vendor/autoload.php';
+
+
+$loop = React\EventLoop\Factory::create();
+
+$callback = function(RequestInterface $request) use ($loop){
+    $input = new ReadableStream();
+
+    $periodicTimer = $loop->addPeriodicTimer(0.5, function () use ($input) {
+        $input->emit('data', array("6\r\nworld\n\r\n"));
```
Comment (clue) [inline examples/streamServer.php]: There's no documentation on what this line is supposed to to. Besides, does this really make sense to ask consumers of this lib to encode this themselves?

File: examples/streamServer.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    });
+
+    $loop->addTimer(3, function () use ($input, $periodicTimer) {
+        $periodicTimer->cancel();
+        $input->emit('data', array("3\r\nend\r\n0\r\n\r\n"));
```
Comment (clue) [inline examples/streamServer.php]: see above

File: examples/streamServer.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        $input->emit('end');
+    });
+
+    $body = new HttpBodyStream($input);
+
+    return new Response(
+        200,
+        array(
+            'Transfer-Encoding' => 'chunked',
```
Comment (clue) [inline examples/streamServer.php]: see above

File: examples/streamServer.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            'Connection' => 'keep-alive'
```
Comment (clue) [inline examples/streamServer.php]: Is this related?

File: src/HttpBodyStream.php [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    {
+        $this->input = $input;
+
+        $this->input->on('data', array($this, 'handleData'));
+        $this->input->on('end', array($this, 'handleEnd'));
+        $this->input->on('error', array($this, 'handleError'));
+        $this->input->on('close', array($this, 'close'));
+    }
+
+    public function handleData($data)
```
Comment (clue) [inline src/HttpBodyStream.php]: Are these really public? See also below

File: src/HttpBodyStream.php [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+    /**
+     *
+     * {@inheritdoc}
+     *
+     * @see \Psr\Http\Message\StreamInterface::getContents()
+     */
+    public function getContents()
+    {
+        echo "test";
```
Comment (clue) [inline src/HttpBodyStream.php]: ?

File: src/HttpServer.php [hunk 7]
```diff
@@ -189,11 +190,23 @@ private function handlePromise(Connection $connection, Promise $promise)
         $promise->then(
             function ($response) use ($connection, $promise){
                 $responseString = RingCentral\Psr7\str(new Response(500));
-
                 if ($response instanceof Response) {
+                    $body = $response->getBody();
+                    if ($body instanceof ReadableStreamInterface) {
+                        $headerResponse = new Response(
```
Comment (clue) [inline src/HttpServer.php]: What is this supposed to do?

**Review by clue — CHANGES_REQUESTED** [review]

**Comment by legionth** [comment]
Hey @clue, have look at this.

File: examples/streamServer.php (hunk 1 above)
Comment (legionth) [reply examples/streamServer.php]: Need a comment here? Is this not readable?

The consumer can write whatever he wants, but it must be valid HTTP.
I could a `Content-Length` example instead, if it makes clearer whats the purpose.

File: src/HttpServer.php (hunk 7 above)
Comment (legionth) [reply src/HttpServer.php]: Create a response object to send the header at the beginning. The body will be streamed.

I found no possibilities to get just the header information as a valid HTTP string to send it to the client.

Any ideas?

File: src/HttpBodyStream.php (hunk 5 above)
Comment (legionth) [reply src/HttpBodyStream.php]: Yes these ar public because, I use them like this `$this->input->on('data', array($this, 'handleData'));
` in the constructor. I could define them as an anonymous function, in the constructor, if you think this would be better.

Comment (legionth) [reply src/HttpBodyStream.php]: Do you see any security issues here?

File: examples/streamServer.php (hunk 1 above)
Comment (clue) [reply examples/streamServer.php]: My point is: I'd rather not ask consumers of this package to learn about the HTTP protocol and chunked transfer encoding etc.

Afaict this can be handled internally, so that consumers do not have to care about this at all :+1:

File: src/HttpServer.php (hunk 7 above)
Comment (clue) [reply src/HttpServer.php]: Maybe I'm missing something obvious, but how about `withBody($emptyBody);` instead?

**Comment by clue** [comment]
What's up with the tests?

File: src/HttpBodyStream.php (hunk 5 above)
Comment (clue) [reply src/HttpBodyStream.php]: I'm not really concerned about *security* here, but more about a stable API (and its documentation). Are these internal only? If so, this should be marked as internal.

File: src/HttpServer.php (hunk 7 above)
Comment (legionth) [reply src/HttpServer.php]: I think you are right. Thank you :+1:

File: src/HttpBodyStream.php (hunk 5 above)
Comment (legionth) [reply src/HttpBodyStream.php]: Ok will do._

**Comment by legionth** [comment]
Added the tests with the last commit. Now all remarks should be handled. Have a look @clue

File: src/ChunkedEncoderStream.php [hunk 8]
```diff
@@ -0,0 +1,143 @@
+<?php
+namespace Legionth\React\Http;
+
+use Evenement\EventEmitter;
+use React\Stream\ReadableStreamInterface;
+use React\Stream\WritableStreamInterface;
+use React\Stream\Util;
+
+class ChunkedEncoderStream extends EventEmitter implements ReadableStreamInterface
```
Comment (clue) [inline src/ChunkedEncoderStream.php:10]: `@internal`?

File: examples/streamServer.php [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    });
+
+    $body = new HttpBodyStream($input);
+
+    return new Response(
+        200,
+        array(
+            'Transfer-Encoding' => 'chunked'
+        ),
+        $body);
```
Comment (clue) [inline examples/streamServer.php]: CS

File: examples/streamServer.php [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+use Legionth\React\Http\HttpBodyStream;
+use React\Socket\Server as Socket;
+use RingCentral\Psr7\Response;
+use Psr\Http\Message\RequestInterface;
+use React\Stream\ReadableStream;
+use Legionth\React\Http\ChunkedEncoderStream;
+
+require __DIR__ . '/../vendor/autoload.php';
+
+
```
Comment (clue) [inline examples/streamServer.php]: CS

File: src/ChunkedEncoderStream.php [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        return $chunkBeginning . $data . "\r\n";
+    }
+
+    /**
+     *
+     * {@inheritdoc}
+     *
+     * @see \React\Stream\ReadableStreamInterface::isReadable()
+     */
+
```
Comment (clue) [inline src/ChunkedEncoderStream.php]: CS and useless docs IMHO? (see also below!)

File: src/ChunkedEncoderStream.php (hunk 8 above)
Comment (clue) [reply src/ChunkedEncoderStream.php:10]: Also, could use some documentation

File: src/ChunkedEncoderStream.php [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        $this->input->on('error', array($this, 'handleError'));
+        $this->input->on('close', array($this, 'close'));
+    }
+
+    /**
+     * @internal
+     */
+    public function handleData($data)
+    {
+        if (is_string($data)) {
```
Comment (clue) [inline src/ChunkedEncoderStream.php]: Why `is_string()`?

File: src/ChunkedEncoderStream.php [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        $this->emit('error', array($e));
+        $this->close();
+    }
+
+    /**
+     * @internal
+     */
+    public function handleEnd($data = null)
+    {
+        if ($data != null && is_string($data)) {
```
Comment (clue) [inline src/ChunkedEncoderStream.php]: Strict comparison and `is_string()` (see above)?

File: src/HttpBodyStream.php [hunk 14]
```diff
@@ -0,0 +1,261 @@
+<?php
+namespace Legionth\React\Http;
+
+use Psr\Http\Message\StreamInterface;
+use React\Stream\ReadableStreamInterface;
+use React\Stream\WritableStreamInterface;
+use Evenement\EventEmitter;
+use React\Stream\Util;
+
+class HttpBodyStream extends EventEmitter implements StreamInterface, ReadableStreamInterface
```
Comment (clue) [inline src/HttpBodyStream.php:17]: Documentation?

File: src/HttpBodyStream.php [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+            $this->close();
+        }
+    }
+
+    /**
+     *
+     * {@inheritdoc}
+     *
+     * @see \Psr\Http\Message\StreamInterface::__toString()
+     */
```
Comment (clue) [inline src/HttpBodyStream.php]: Useless IMHO? (see also below!)

File: src/HttpServer.php [hunk 16]
```diff
@@ -189,11 +190,18 @@ private function handlePromise(Connection $connection, Promise $promise)
         $promise->then(
             function ($response) use ($connection, $promise){
                 $responseString = RingCentral\Psr7\str(new Response(500));
-
                 if ($response instanceof Response) {
+                    $body = $response->getBody();
+                    if ($body instanceof ReadableStreamInterface) {
+                        $emptyBody =RingCentral\Psr7\stream_for('');
```
Comment (clue) [inline src/HttpServer.php]: CS and some comments would make sense?

**Review by clue — CHANGES_REQUESTED** [review]
Nice! :+1:

Loads of minor remarks, otherwise LGTM :shipit:

**Comment by legionth** [comment]
I think I dealed with the every minor remarks you had @clue ? What do you say?

File: src/ChunkedEncoderStream.php [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+
+use Evenement\EventEmitter;
+use React\Stream\ReadableStreamInterface;
+use React\Stream\WritableStreamInterface;
+use React\Stream\Util;
+
+/**
+ * Stream incoming strings as a HTTP chunked encoded string
+ */
+class ChunkedEncoderStream extends EventEmitter implements ReadableStreamInterface
```
Comment (clue) [inline src/ChunkedEncoderStream.php:10]: Should this be `@internal`?

File: src/ChunkedEncoderStream.php [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    /**
+     * @param string $data - string to be transformed in an valid
+     *                       HTTP encoded chunk string
+     * @return string
+     */
+    private function createChunk($data)
+    {
+        if (is_string($data)) {
```
Comment (clue) [inline src/ChunkedEncoderStream.php]: Docs suggest this is always a string?

File: src/HttpBodyStream.php [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    public function write($string)
+    {
+        return;
+    }
+
+    public function read($length)
+    {
+        // TODO: Auto-generated method stub
```
Comment (clue) [inline src/HttpBodyStream.php]: ? (see also above and below)

**Review by clue — CHANGES_REQUESTED** [review]

**Comment by legionth** [comment]
Ping @clue . I created a new class `StreamedResponse`. I'm not sure about this class, maybe a factory or something else into this direction would be better. What do you think?

**Comment by clue** [comment]
Not sure I get how `StreamedResponse` has any advantaged over the previous solution? Also, how would I set other response parameters, such as HTTP status codes, headers etc. etc.

Also, how does this integrate with the existing middleware stack? I suppose the previous approach would make integration with existing middlewares less of an issue?

**Comment by legionth** [comment]
Well it is still just a repsonse object. And you can set status code and and headers etc. via the constructor.
The advantages are you don't need to create a `ChunkedEncoderStream` or `HttpBodyStream`, you also don't need to add the header of 'Transfer-Encoding: chunked' by yourself.

I see no problems with the current middlewarestack because this is still a `Response` object.

**Comment by legionth** [comment]
Hey @clue . I moved the `ChunkedEncoderStream` to the `HttpBodyStream` and the `HttpServer` will now automatically  add a 'Transfer-Encoding: chunked' if the user hasn't defined it.

What do you think about this?

File: README.md [hunk 20]
```diff
@@ -187,6 +188,37 @@ This little example should show how you can use the middlwares e.g. to check or
...
+        $body
+    );
+}
+```
+
+The `HttpServer` will use the emitted data from the `ReadableStream` to send this data directly to the client.
+
+You can add an the optional encoder as second parameter of the constructor of the `HttpBodyStream`. If none is set the default will the be`ChunkedEncoderStream` an 
+encoder which will send HTTP encoded chunks to the client.
+The `HttpServer` will automaticly add a 'Transfer-Encoding: chunked' line to your header, if you use the `HttpBodyStream` with the default `ChunkedEncoderStream`.
```
Comment (clue) [inline README.md]: This is a fairly good documentation about the *what*, what's more important to consumers of this library is probably the *why*

File: src/ChunkedEncoderStream.php (hunk 17 above)
Comment (clue) [reply src/ChunkedEncoderStream.php:10]: What "incoming strings"? Perhaps something like "Wraps any readable stream which may emit number of data chunks as strings", also some documentation about the *why*?

File: src/HttpBodyStream.php [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+
+use Psr\Http\Message\StreamInterface;
+use React\Stream\ReadableStreamInterface;
+use React\Stream\WritableStreamInterface;
+use Evenement\EventEmitter;
+use React\Stream\Util;
+
+/**
+ * Uses a StreamInterface from PSR-7 and a ReadableStreamInterface from ReactPHP
+ * to use PSR-7 objects and use ReactPHP streams
```
Comment (clue) [inline src/HttpBodyStream.php:12]: *why*?

File: src/HttpServer.php [hunk 22]
```diff
@@ -205,6 +220,38 @@ function () use ($connection) {
     }
 
     /**
+     * Check if a ChunkEncoderStream is setted as encoder in the bodey stream
+     * @param ReadableStreamInterface $body - body to be checked for chunked encoding
+     * @return boolean
+     */
+    public function hasChunkedEncoderStream($body)
```
Comment (clue) [inline src/HttpServer.php]: This method looks kind of odd to me, perhaps because it violates SRP?

File: src/HttpBodyStream.php [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+    private $closed = false;
+    private $encoder;
+
+    public function __construct(ReadableStreamInterface $input, ReadableStreamInterface $encoder = null)
+    {
+        $this->input = $input;
+
+        if ($encoder === null) {
+            $encoder = new ChunkedEncoderStream($this->input);
+        }
```
Comment (clue) [inline src/HttpBodyStream.php]: This makes sense from a SOLID standpoint (DI), but it looks kind of odd from a consumer perspective. Perhaps the constructor could explain the *why*?

File: src/HttpBodyStream.php [hunk 24]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    public function __toString()
+    {
+        return '';
+    }
+
+    public function detach()
+    {}
```
Comment (clue) [inline src/HttpBodyStream.php]: Not sure about this (and the following below). Also, any documentation?

**Review by clue — CHANGES_REQUESTED** [review]

File: src/HttpBodyStream.php (hunk 24 above)
Comment (legionth) [reply src/HttpBodyStream.php]: I could always throw a `BadMethodCallException`?

These are the methods from `StreamInterface` so they aren't really needed anywhere. I could write that, but I don't know if that would help.

File: src/HttpServer.php (hunk 22 above)
Comment (legionth) [reply src/HttpServer.php]: Maybe but have currently no better ideas. :zipper_mouth_face:

**Comment by legionth** [comment]
Hey @clue , I moved the chunked enccoding to the `HttpServer`, the encoding will now always be chunked encoding. I also added some documentation.

File: README.md [hunk 25]
```diff
@@ -187,6 +188,36 @@ This little example should show how you can use the middlwares e.g. to check or
...
+    return new Response(
+        200,
+        array(),
+        $body
+    );
+}
+```
+
+The `HttpServer` will use the emitted data from the `ReadableStream` to send this data directly to the client.
+If you use the `HttpBodyStream` the whole communication to the client will be HTTP chunk encoded, other values set for `Transfer-Encoding` will be ignored.
```
Comment (clue) [inline README.md]: "will be HTTP chunk encoded", what does that mean? :)

File: README.md [hunk 26]
```diff
@@ -187,6 +188,36 @@ This little example should show how you can use the middlwares e.g. to check or
...
+
+Big amount of data can easily be handled and doesn't need to be saved temporarily, because it will be send directly to the client.
```
Comment (clue) [inline README.md]: typos

File: README.md [hunk 27]
```diff
@@ -187,6 +188,36 @@ This little example should show how you can use the middlwares e.g. to check or
 
 Checkout the `examples/middleware` how to add multiple middlewares.
 
+### Streaming body data
+
+If you have operations that need time to be computed(of course non-blocking ;-)), you can stream the intermediary results directly to the client,
```
Comment (clue) [inline README.md]: "need time to be computed(of course non-blocking ;-))"

File: src/ChunkedEncoderStream.php [hunk 28]
```diff
@@ -0,0 +1,113 @@
...
+
+    /**
+     * Will emit the given string in a chunked encoded string
+     * @param string $data - string to be tranfsformed into an HTTP chunked encoded string
+     */
+    public function handleData($data)
+    {
+        $completeChunk = $this->createChunk($data);
+
+        $this->emit('data', array($completeChunk));
```
Comment (clue) [inline src/ChunkedEncoderStream.php]: data could be an empty string here

File: src/ChunkedEncoderStream.php [hunk 29]
```diff
@@ -0,0 +1,113 @@
...
+
+    /**
+     * @param string $data - string to be transformed in an valid
+     *                       HTTP encoded chunk string
+     * @return string
+     */
+    private function createChunk($data)
+    {
+        $byteSize = strlen($data);
+        $chunkBeginning = $byteSize . "\r\n";
```
Comment (clue) [inline src/ChunkedEncoderStream.php]: nope :smiling_imp:

File: src/HttpBodyStream.php [hunk 30]
```diff
@@ -0,0 +1,185 @@
...
+        $this->input->on('data', array($this, 'handleData'));
+        $this->input->on('end', array($this, 'handleEnd'));
+        $this->input->on('error', array($this, 'handleError'));
+        $this->input->on('close', array($this, 'close'));
+    }
+
+    public function getEncoder()
+    {
+        return $this->encoder;
+    }
```
Comment (clue) [inline src/HttpBodyStream.php]: ?

File: src/HttpBodyStream.php [hunk 31]
```diff
@@ -0,0 +1,185 @@
...
+
+    /**
+     * Emits the data
+     * @param string $data - data to be emitted
+     */
+    public function handleData($data)
```
Comment (clue) [inline src/HttpBodyStream.php]: `@internal` (see also below)

File: src/HttpServer.php [hunk 32]
```diff
@@ -179,21 +180,36 @@ private function executeMiddlewareChain(array $middlewareChain, RequestInterface
...
         $promise->then(
-            function ($response) use ($connection, $promise){
+            function ($response) use ($connection, $promise, $that){
                 $responseString = RingCentral\Psr7\str(new Response(500));
-
                 if ($response instanceof Response) {
+                    $body = $response->getBody();
+                    if ($body instanceof ReadableStreamInterface) {
+                        // reset Transfer-Encoding header and set always chunked encoding
+                        $response = $response->withoutHeader('Transfer-Encoding');
```
Comment (clue) [inline src/HttpServer.php]: not needed

File: src/HttpServer.php [hunk 33]
```diff
@@ -179,21 +180,36 @@ private function executeMiddlewareChain(array $middlewareChain, RequestInterface
...
+                        $response = $response->withHeader('Transfer-Encoding', 'chunked');
+                        // Send the header first without the body,
+                        // the body will be streamed
+                        $emptyBody = RingCentral\Psr7\stream_for('');
+                        $headerResponse = $response->withBody($emptyBody);
```
Comment (clue) [inline src/HttpServer.php]: `$response` vs `$headerResponse`?

**Review by clue — CHANGES_REQUESTED** [review]

**Comment by legionth** [comment]
Ping @clue . The last commit should fix the latest remarks you had.

File: src/ChunkedEncoderStream.php [hunk 34]
```diff
@@ -0,0 +1,118 @@
...
+
+use Evenement\EventEmitter;
+use React\Stream\ReadableStreamInterface;
+use React\Stream\WritableStreamInterface;
+use React\Stream\Util;
+
+/**
+ * Wraps any readable stream which will emit the data as HTTP chunks
+ */
+class ChunkedEncoderStream extends EventEmitter implements ReadableStreamInterface
```
Comment (clue) [inline src/ChunkedEncoderStream.php:10]: `@internal`?

File: README.md (hunk 25 above)
Comment (clue) [reply README.md]: ?

File: examples/streamServer.php [hunk 35]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+use Legionth\React\Http\HttpServer;
+use Legionth\React\Http\HttpBodyStream;
+use React\Socket\Server as Socket;
+use RingCentral\Psr7\Response;
+use Psr\Http\Message\RequestInterface;
+use React\Stream\ReadableStream;
+use Legionth\React\Http\ChunkedEncoderStream;
```
Comment (clue) [inline examples/streamServer.php]: leftover?

File: src/ChunkedEncoderStream.php [hunk 36]
```diff
@@ -0,0 +1,118 @@
...
+        $this->input->on('close', array($this, 'close'));
+    }
+
+    /**
+     * Will emit the given string in a chunked encoded string
+     * @param string $data - string to be tranfsformed into an HTTP chunked encoded string
+     */
+    public function handleData($data)
+    {
+        if ($data == '') {
```
Comment (clue) [inline src/ChunkedEncoderStream.php]: strict comparison

**Review by clue — CHANGES_REQUESTED** [review]

**Comment by legionth** [comment]
Ping @clue. Fixed the latest remarks.

**Review by clue — APPROVED** [review]
LGTM :shipit: :tada:
